### DISCUSSION TITLE
How to create a map via API with datasets in it

### DISCUSSION THREAD

**Discussion by nicokant** [GeoNode/geonode#12139] [Q&A]
I'm trying to migrate some maps from a geonode 3 to a geonode 4 instance, I see that there is a REST API available to create maps in geonode 4 and it's used by the mapstore client, but it's not clear how to use it since it expects lots of nested data and the REST documentation is not explicit about it. Could you provide an example of a valid request/template request of the required parameters?

**Comment by giohappy** [comment]
@nicokant The maps API expects two different types of information:

 - the list of `maplayers` that are referenced by the map -> this is only needed for metadata purposes, and for rendering the map thumbnail
 - the `data` JSON, which is opaque to GeoNode itself -> it contains the map configuration for the client

You can see the two blocks of information sent when a new map is created from the client:
![image](https://github.com/GeoNode/geonode/assets/571129/c30b4cab-86f8-4915-80d2-a1d36a58333a)

The connection between the layers configuration inside the `data` object and `maplayers` is provided by the `msId` property, that is generated by the client when the map is created.

The second is the internal representation of the [map configuration from MapStore](https://docs.mapstore.geosolutionsgroup.com/en/latest/developer-guide/maps-configuration/). The GeoNode backend ignores it. It just take it and saves it in the DB. So, for the moment, the programmatic creation of a map as you view it in the map viewer is not supported.

**Reply by ridoo** [reply]
Did not spot this so far .. interesting part. I guess, [this is the relevant location](https://github.com/GeoNode/geonode/blob/6ce7f866abe00c44f37ec36ed5839eba0d7058fa/geonode/base/api/serializers.py#L707-L715) where it is stored as blob in the database:

```py
    def to_internal_value(self, data):
        if isinstance(data, str):
            data = json.loads(data)
        if "data" in data:
            data["blob"] = data.pop("data")
        if isinstance(data, QueryDict):
            data = data.dict()
        data = super(ResourceBaseSerializer, self).to_internal_value(data)
        return data
```

As this is theoretically available for every resource type, are there any other resources or use cases you can think of where it makes sense to store arbitrary `data` as blobs?

**Reply by nicokant** [reply]
It kind makes sense, but it also makes it very hard to interact with that structure without executing any javascript code in a full mapstore environment.

How does django migrations handle the changes to the maps model to this "frontend-only" structure from v3 to v4? Is there a way or the maps must be deleted and recreated manually while performing this kind of upgrade?

**Reply by ridoo** [reply]
@nicokant as far I can see, storing map configuration as blob into the `data` field is a new feature since GeoNode `v4.0`. In version `v3` the [map config was a more plain](https://github.com/GeoNode/geonode/blob/7c25d148950accaf02bf7e5b4fd62bfca4027627/geonode/maps/models.py#L62-L77).

There is an [initial Django migration](https://github.com/GeoNode/geonode/blob/1970a810ecc72f893c432c9d586d16eaa2ee3656/geonode/maps/migrations/0030_auto_20210506_0836.py#L19) for the blob field where [config got moved](https://github.com/GeoNode/geonode/blob/1970a810ecc72f893c432c9d586d16eaa2ee3656/geonode/maps/migrations/0031_auto_20210615_0717.py#L5-L9) from some older [`maps_mapdata` model](https://github.com/GeoNode/geonode/blob/1970a810ecc72f893c432c9d586d16eaa2ee3656/geonode/maps/migrations/0030_auto_20210506_0836.py#L15-L21). I cannot get the right line, but it seems `mapdata` had been an intermediary model. ATM, I cannot find any migration where the old map config (including `center_x`, `center_y`) got converted to a MapStore config JSON.

Maybe you just find out by applying the migrations on a test setup?

**Reply by nicokant** [reply]
Thank you, I tried searching the source too but nothing seems to pop up. I noticed that there is an example "dummy" structure for maps in the api tests (https://github.com/GeoNode/geonode/blob/master/geonode/maps/api/tests.py), is this the minimum structure that the api supports to create a map?

> Maybe you just find out by applying the migrations on a test setup?
It could be a good idea, thank you :)

**Reply by ridoo** [reply]
Not sure if this is minimal but this should be the structure who a map configuration looks like. More information about how the map config should look like [can be found in here](https://github.com/geosolutions-it/MapStore2/blob/master/docs/developer-guide/maps-configuration.md#map-options) (or have a look at the official [deployed MapStore docs](https://docs.mapstore.geosolutionsgroup.com/en/v2023.02.02/developer-guide/maps-configuration/))

**Reply by nicokant** [reply]
> There is an [initial Django migration](https://github.com/GeoNode/geonode/blob/1970a810ecc72f893c432c9d586d16eaa2ee3656/geonode/maps/migrations/0030_auto_20210506_0836.py#L19) for the blob field where [config got moved](https://github.com/GeoNode/geonode/blob/1970a810ecc72f893c432c9d586d16eaa2ee3656/geonode/maps/migrations/0031_auto_20210615_0717.py#L5-L9) from some older [maps_mapdata model](https://github.com/GeoNode/geonode/blob/1970a810ecc72f893c432c9d586d16eaa2ee3656/geonode/maps/migrations/0030_auto_20210506_0836.py#L15-L21). I cannot get the right line, but it seems mapdata had been an intermediary model. ATM, I cannot find any migration where the old map config (including center_x, center_y) got converted to a MapStore config JSON.

I haven't tested yet, but I'm quite sure it won't work, for what I understood from the source:
- There was a temporary django model that implemented the logic of transformation from v3 maps to v4 maps, but the logic itself was probably in python
- the migration 30 actually creates the model in the database, but there is no source code for that model
- the migration 31 performs a SQL query to move the blob data and then it deletes the model in the database

Probably this was initially a set of multiple commits, but then were squashed and the model that was created and deleted in different commits disappeared from code. I'm not really sure of that, this is mostly a deduction but it worth investigating because it would mean that migration from v3 to v4 is not working
